Restart sourcekitd and clangd after they have crashed

ahoppen opened the pull request:
Recover from crashes in `sourcekitd` and `clangd` by transparently restarting them from SourceKit-LSP.

ahoppen commented:
https://github.com/apple/swift/pull/29558
@swift-ci Please test

ahoppen commented on Sources/SourceKit/clangd/ClangLanguageServer.swift:
```diff
@@ -17,29 +17,136 @@ import LSPLogging
+    process.terminationHandler = { [weak self] process in
+      log("clangd exited: \(process.terminationReason) \(process.terminationStatus)")
+      connection.close()
+      if process.terminationStatus != 0 {
```
We’re already doing it that way, I just copied the code over from here. AFAIK it never hurts to close the connection, if it’s already closed, its a no-op. So it’s better to just close it.
https://github.com/apple/sourcekit-lsp/blob/master/Sources/SourceKit/clangd/ClangLanguageServer.swift#L238

DavidGoldman replied:
Ah, I meant the part below, where `clangd` is restarted. Maybe you should double check the `state` or see if it was shutting down or equivalent?

ahoppen replied:
What do you want to check. `state` is always either `connected` or `connectionInterrupted` in `ClangLanguageServerShim ` since we never shut the connection to `clangd` down. What kind of check are you proposing?

DavidGoldman replied:
We should probably handle the exit notification and properly shut it down, perhaps that should be in a follow up change though

ahoppen replied:
If we send an exit notification, I think it is safe to assume that `clangd` will exit with exit code 0 which means we won’t restart it (since we check for `process.terminationStatus != 0`). I think we are misunderstanding each other though.

DavidGoldman replied:
Ah I was thinking about signals, but I don't think SourceKit-LSP has any signal handlers right now so it should crash.

ahoppen replied:
Ah, OK. No, there’s no signal handlers so we should be fine.

ahoppen commented on Sources/SourceKit/clangd/ClangLanguageServer.swift:
```diff
@@ -17,29 +17,136 @@ import LSPLogging
+      self.reopenDocuments()
+      self.state = .connected
+    } catch {
+      log("Failed to restart clangd after a crash, retrying in 5 seconds: \(error)", level: .error)
```
On second thought, I think we probably shouldn’t try to restart the server if this fails anyway. It means that we were either unable to launch the process or that `clangd` returned an error from the `InitializeRequest`. In both cases, it’s unlikely that another relaunch will solve the issue.

ahoppen commented on Sources/SourceKit/SourceKitServer.swift:
```diff
@@ -193,7 +223,16 @@ public final class SourceKitServer: LanguageServer {
+        switch language {
+        case .c, .cpp, .objective_c, .objective_cpp:
+          // Re-open all C-based languages if clangd crashed
+          self.reopenDocuments(for: [.c, .cpp, .objective_c, .objective_cpp])
```
That’s a good idea, much cleaner than hard coding the languages. Thanks!

ahoppen commented on Sources/SourceKit/SourceKitServer.swift:
```diff
@@ -185,6 +185,36 @@ public final class SourceKitServer: LanguageServer {
     return nil
   }
 
+  func reopenDocuments(for languages: Set<Language>) {
```
Probably it should. Thanks!

ahoppen commented on Sources/SourceKit/SourceKitServer.swift:
```diff
@@ -418,13 +458,17 @@ extension SourceKitServer {
 
-    let textDocument = note.params.textDocument
+    let textDocument = note.textDocument
     workspace.buildSettings.registerForChangeNotifications(for: textDocument.uri)
```
I guess, I should close the document the same way as I open it, by sending a `DidCloseTextDocumentNotification`. That should take care of unregistering it as well.

ahoppen replied:
Actually on second thought, we should probably just not re-register for change notifications. As far as `sourcekit-lsp` is concerned, the files are still open, there has just been a hick-up with the underlying functionality. As far as the build system is concerned, we can still receive updated build settings and they will be used when `sourcekitd` restarts (see the comment about not resetting the `buildSettingsByFile`). 

What do you think?

DavidGoldman replied:
Yeah, that makes sense, but we'll need to make sure that each `ToolchainLanguageServer` correctly receives the new build settings or caches it (including clangd)

ahoppen replied:
Since the restarted clang language server is being initialized with the same build settings as before (https://github.com/apple/sourcekit-lsp/blob/master/Sources/SourceKit/clangd/ClangLanguageServer.swift#L211) and `SwiftLanguageServer` sends the build settings on document open (https://github.com/apple/sourcekit-lsp/blob/master/Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift#L254), we should be safe here. Might be worth to add a test case for though.

DavidGoldman replied:
Ah right as long as long as the `ClangLanguageServerShim` receives the new open document request it should be OK. Yeah a test for this seems good

ahoppen replied:
Added a test case that verifies that the build settings are recovered after a crash.

ahoppen commented on Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift:
```diff
@@ -113,6 +130,27 @@ extension SwiftLanguageServer {
+      if let error = notification.error, error.code == .connectionInterrupted {
+        self.state = .connectionInterrupted
+
+        // We don't have any open documents anymore after sourcekitd crashed.
```
I don’t think so. Even though `sourcekitd` crashed, the build settings should still be correct, so we should be able to keep them around. Do you agree?

DavidGoldman replied:
Yeah that seems reasonable although if the re-open fails you might never wipe them

ahoppen replied:
Well, even if the re-open fails, the build settings are correct. They are just no longer used. I don’t see a reason to wipe them ever.

DavidGoldman replied:
Yeah, it shouldn't be a big deal but in theory if it crashes with a lot of documents open it will save all of the build settings. But I don't think it's a major concern

ahoppen replied:
Who will save the build settings?

Also: What’s the problem with saving the build settings anyway. They are still correct.

DavidGoldman replied:
Well `SwiftLanguageServer` does, imagine a case:
- SourceKitServer has the file open
- SwiftLanguageServer has the file open and saved build settings
- SwiftLanguageServer crashes, reopen fails
- Potential build system changes?

But yeah I don't think this is really an issue

ahoppen replied:
I assume by `SwiftLanguageServer`, you mean `sourcekitd` (the `SwiftLanguageServer` never writes anything to disk).

And if `sourcekitd` fails to re-open the file, well, there’s no way we would be able to tell it about the new build settings anyway. Keep in mind that `buildSettingsByFile` keeps being updated even if `sourcekitd` has crashed.

DavidGoldman replied:
Ah I see, for some reason I thought the `buildSettingsByFile` wouldn't be updated anymore

ahoppen commented:
https://github.com/apple/swift/pull/29558
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

ahoppen commented:
> It might also make sense to swap `FileBuildSettings` to be a class since it's commonly shared/cached and immutable, not sure.

I don’t think so, we explicitly want the value semantics. Plus it’s not immutable, both `compilerArguments` and `workingDirectory` are `var`s.

benlangmuir commented on Sources/SourceKit/sourcekitd/SwiftSourceKitFramework.swift:
```diff
@@ -304,6 +304,7 @@ struct sourcekitd_keys {
 
 struct sourcekitd_requests {
 
+  let crash_with_exit: sourcekitd_uid_t
```
Nitpick: `crash_exit` the name should match the UID

benlangmuir commented on Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift:
```diff
@@ -131,6 +148,27 @@ extension SwiftLanguageServer {
       let notification = SKResponse(notification, sourcekitd: self.sourcekitd)
 
+      if notification.value?[self.keys.notification] == self.values.notification_sema_enabled {
+        self.state = .connected
```
The notification handler queue is not related to `queue` for this object, so to access `state` we need to synchronize. Similarly, `addStateChangeHandler` needs to synchronize the same way.

ahoppen replied:
As I understood it, `queue` is used to synchronise the requests and responses from `sourcekitd`. Setting the state does not fall into this category, so I did not synchronise it on `queue`. 

Do you want to extend the definition of `queue`? I don’t think that `state` really needs to be synchronised, but it wouldn’t hurt either.

benlangmuir replied:
The comment on `queue` is misleading and should be updated.  We need to protect shared access to mutable state - e.g. we're doing this for `currentDiagnostics` already.

ahoppen replied:
OK, in that case, it makes total sense to guard the access to state by `queue`.

benlangmuir commented on Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift:
```diff
@@ -192,7 +230,14 @@ extension SwiftLanguageServer {
+  }
+
+  /// Tell sourcekitd to crash itself. For testing purposes only.
+  public func crash() {
```
Naming: `_crash` since it's public for testing

benlangmuir commented on Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift:
```diff
@@ -131,6 +148,27 @@ extension SwiftLanguageServer {
+
+        // We don't have any open documents anymore after sourcekitd crashed.
+        // Reset the document manager to reflect that.
+        self.documentManager = DocumentManager()
```
This mutation also needs synchronization.  It's kind of suspicious that you're doing this *after* you call reopenDocuments - couldn't that repopulate it at the wrong time?

ahoppen replied:
> This mutation also needs synchronization.

Good point. I wrapped it in `queue.async`.

> It's kind of suspicious that you're doing this after you call reopenDocuments - couldn't that repopulate it at the wrong time?

No, that ordering is correct. When `sourcekitd` crashes, `state` is `.connected`. Hence the second `if self.state == .connectionInterrupted` is skipped and the third `if let error = notification.error, error.code == .connectionInterrupted` sets `state = .connectionInterrupted`, because we receive a notification with `error.code == .connectionInterrupted`. 

When the next notification comes in, the first `if self.state == .connectionInterrupted` is taken, which then sets `state = .semanticFunctionalityDisabled`.

Finally, when semantic functionality is restored, we get a notification that satisfies the first `if notification.value?[self.keys.notification] == self.values.notification_sema_enabled`, setting `state` back to `.connected`.

benlangmuir commented on Sources/SourceKit/clangd/ClangLanguageServer.swift:
```diff
@@ -24,22 +24,125 @@ final class ClangLanguageServerShim: ToolchainLanguageServer {
+      connection.close()
+      if process.terminationStatus != 0 {
+        if let self = self {
+          self.state = .connectionInterrupted
```
Similar synchronization issues to the Swift side.

ahoppen replied:
Same comment as on the Swift side: 

As I understood it, queue is used to synchronise the requests and responses from `clangd`. Setting the state does not fall into this category, so I did not synchronise it on queue.

Do you want to extend the definition of queue? I don’t think that state really needs to be synchronised, but it wouldn’t hurt either.

benlangmuir replied:
> I don’t think that state really needs to be synchronised, but it wouldn’t hurt either.

Why not? It's being mutated and read concurrently.

ahoppen replied:
I can’t think of a race condition that could happen right now. But it definitely doesn’t hurt to guard it.

benlangmuir commented on Sources/SourceKit/clangd/ClangLanguageServer.swift:
```diff
@@ -24,22 +24,125 @@ final class ClangLanguageServerShim: ToolchainLanguageServer {
+      if process.terminationStatus != 0 {
+        if let self = self {
+          self.state = .connectionInterrupted
+          self.restartClangd()
```
I'm concerned that this could crash loop. For sourcekitd, XPC prevents us from re-spawning too fast in a loop. We should implement something here to delay the relaunch if we crash loop (e.g. after the second crash in a short time start delaying a few seconds before trying again).

DavidGoldman replied:
FWIW VSCode will restart an LSP for up to 5 times within a 3 minute interval. Really depends upon the type of crash - most of which I've seen are persistent or related to a single document.

Might make sense to do something similar here

ahoppen replied:
Good point. I have added some super simple crash-loop prevention code: If `clangd` has crashed in the last 30 seconds, delay any further restart by 10 seconds. I think this should be sufficient to prevent us from going wild im crash loop scenarios. I also added a test case that checks that `clangd` isn’t restarted twice in a row within 5 seconds.

ahoppen commented on Sources/SourceKit/SourceKitServer.swift:
```diff
@@ -185,6 +185,36 @@ public final class SourceKitServer: LanguageServer {
   }
 
+  func reopenDocuments(for languageService: ToolchainLanguageServer) {
+    queue.sync {
```
Good question. I don’t think there is one. I changed it to `async`.

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@benlangmuir I have aded https://github.com/apple/sourcekit-lsp/pull/219/commits/2b0cbfd8d7854a633f15deb9ec5b264823f44734 which guards access to `state` and `stateChangeHandlers` by `queue`. Could you take a look at it?

@swift-ci Please test

ahoppen commented:
@swift-ci Please test

benlangmuir commented on Tests/SourceKitTests/CrashRecoveryTests.swift:
```diff
@@ -0,0 +1,269 @@
+    // Do a sanity check and verify that we get the expected result from a hover response before crashing clangd.
+
+    let expectedHoverContent = """
+      Declared in global namespace
```
This text seems to vary between recent clangd versions - I tried this test locally and it failed because "loc" wasn't included at the bottom.  We generally try to keep our tests passing with both the latest toolchain and the most recent _released_ toolchain.  Let's change this only compare the range, or if you really want to check the markup content, just check a prefix like "Declared in".

ahoppen commented:
Hi @benlangmuir, thanks for running the thread sanitizer. Looks like I missed a couple data races, but things should be fixed now. I also changed crash recovery tests to only verify that the range is correct. I think that should suffice for the sanity check that we have correctly re-opened the document.

I have added then new changes as two new commits for an easier review. I’d like to  squash them all together before merging the PR though.

@swift-ci Please test

benlangmuir commented:
Please regenerate linuxmain; otherwise LGTM.  Thanks!

ahoppen commented:
Squashed and regenerated `LinuxMain.swift`

@swift-ci Please test

ahoppen commented:
Test failure seems to have been unrelated

@swift-ci Please test

benlangmuir commented:
> CrashRecoveryTests.testSourcekitdCrashRecovery : API violation - multiple calls made to XCTestExpectation.fulfill() for clangd crashed.

ahoppen commented:
`assertForOverFulfill` seems to not be working on Linux's `XCTestExpectation`. I have filed [SR-12575](https://bugs.swift.org/browse/SR-12575) for it. I am now manually keeping track if `fulfill()` has already been called on the expectation.

@swift-ci Please test

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

shahmishal commented:
The Swift project moved the default branch to `main` and deleted `master` branch, so GitHub automatically closed the PR. Please re-create the pull request with `main` branch.

More detail about the branch update - https://forums.swift.org/t/updating-branch-names/40412
